projects/DifferentialDrive.md

```diff
@@ -0,0 +1,655 @@
+
+Here, we’re calculating the feedforward voltage for each wheel based on the target speed. This helps to ensure that the motors get an initial boost that’s closer to what they need.
+
+Next, we add the PID control:
+
+```java
+    final double leftOutput = 
+      leftPIDController.calculate(leftEncoder.getVelocity(), 
+      leftSpeed * DriveConstants.MAX_SPEED);
+    final double rightOutput = 
+      rightPIDController.calculate(rightEncoder.getVelocity(), 
+      rightSpeed * DriveConstants.MAX_SPEED);
+```
+
+In this part, the PID controllers calculate how much they need to adjust the motor speed to correct any errors. They do this by comparing the actual speed (from the encoders) with the target speed.
+
+Finally, we combine the outputs and send them to the motors:
+
+```java
+      double leftVoltage = leftOutput + leftFeedforward;
+      double rightVoltage = rightOutput + rightFeedforward;
+
+      leftLeader.setVoltage(leftVoltage);
+      rightLeader.setVoltage(rightVoltage);
+```
+
+Here, the final motor voltage is calculated by adding the PID and feedforward outputs together. These voltages are then sent to the motors, controlling the movement of the robot.
+
+### Unit Testing & System Checks
+
+#### Unit Tests
+
+Think of unit tests as a way to check that each small piece of your code—like a motor or sensor—is functioning properly on its own before you start putting them all together. It's much easier to fix problems when they’re isolated to a single component than when they’re all tangled up in the full system.
+
+**Make sure to check out this [guide](https://docs.wpilib.org/en/stable/docs/software/wpilib-tools/robot-simulation/unit-testing.html) from WPILib on how to setup unit tests.**
+
+Let’s talk about the tests for our drivetrain. We'll start it by creating a `DriveTest.java` file in the robot folder of the tests folder.
+
+First, we need to set things up before we run any tests. This is done in the `setup()` method:
+
+```java
+  Drive drive;
+  CANSparkMax spark;
+
+  @BeforeEach
+  public void setup() {
+    setupTests();
+    drive = new Drive();
+  }
+```
+
+Next, we will need to reset our drive between every tests so we are able to experiment different things at once.
+
+```java
+  @AfterEach
+  void destroy() throws Exception {
+    reset(drive);
+  }
+```
+
+Once everything is set up, you can make tests with the `@Test` annotation. Here's a quick example of testing if our pose is changing as we drive. 
+
+The `fastForward` allows time to pass to give our drive time to move. 
+
+Then we call our `getPose()` method and compare it to a new Pose2d, which returns a value of 0. So this test ensures that whenever our drive method is called, the robot doesn't stay at the Pose2d but actually moves.
+
+```java
+  @Test
+  public void testPose() {
+    drive.drive(0.8, 0.8);
+    fastForward();
+
+    assertNotEquals(new Pose2d(), drive.getPose());
```

Comment:
i feel like you should test that it goes in the right direction, not just that it moves. Ideally, you could even check that it goes a reasonable distance. You can multiply the drive speed by the time (you can call fastforward with the seconds you want it to move), and check that the position is in some range around that point